cloudinit/sources/DataSourceOVF.py

```diff
@@ -152,14 +152,25 @@ def _get_data(self):
                     product_marker, os.path.join(self.paths.cloud_dir, 'data'))
                 special_customization = product_marker and not hasmarkerfile
                 customscript = self._vmware_cust_conf.custom_script_name
-                custScriptConfig = get_tools_config(
-                    CONFGROUPNAME_GUESTCUSTOMIZATION,
-                    GUESTCUSTOMIZATION_ENABLE_CUST_SCRIPTS,
-                    "false")
-                if custScriptConfig.lower() != "true":
-                    # Update the customization status if there is a
-                    # custom script is disabled
-                    if special_customization and customscript:
+
+                # In case there is a custom script, check whether VMware
+                # Tools configuration allow the custom script to run.
+                if special_customization and customscript:
+                    defVal = "false"
+                    if self._vmware_cust_conf.default_run_post_script:
```

Comment:
@xiaofengw-vmware, sorry to rehash this earlier discussion, but I need a bit of clarification on the deployment 'flow' and a vmware user launching a vm in vCloud (not vmware admin) would follow in vCloud product.  

I have a couple of questions about this architecture that hopefully you can clarift, because this doesn't seem to meet this concerns @smoser mentioned on the previous PR #426. 

Please correct any misunderstanding I have about vCenter, vSphere, or vCloud products. If something sounds incorrect to you, it probably is due to my lack of understanding of VMware products.

1. Is the intent of this branch is force the post-cloud-init-customization script to run of VM Tools configuration does not provide a value of true for "enable-custom-scripts" from the utility "vmware-toolbox-cmd". (I see the following [doc reference about enable-custom-scripts](https://kb.vmware.com/s/article/74880))

2. Does a user launching each VM in vCloud have the opportunity to override this config option to "false" on the command line and/or UI during each vm creation of their images?  What does that look like in vCloud/vSphere world? 

3. I believe [the concern @smoser raised ](https://github.com/canonical/cloud-init/pull/426#issuecomment-644772401) may have been getting at is the following:
   * a given vCenter/vSphere or vCloud admin chooses a to default post-install script behavior to run on all VMs launched on that vCenter or vCloud deployment through VMTools configuration or through vCloud default_run_post_script setting
   * it would be ideal if during each vm creation, the user launching an image/vm could specify an override in cloud-init which takes precedence over both vm tools config and vCloud default_run_post_script setting.

In this case we consider both VM Tools config and vCloud default_run_post_script to be "cloud platform default behavior"  but we need a way for a cloud-init consumer to disable such behavior if desired.


One way to do that is to provide datasource-configuration options for DataSourceOVF which are read from the class attribute `self.ds_cfg` and is already being done for the flag [vmware_cust_file_max_wait](https://github.com/canonical/cloud-init/blob/master/cloudinit/sources/DataSourceOVF.py#L345-L364) . This would allow a customer to create a custom image that would override any default script opinions that a vCenter|vSphere/vCloud platform had chosen.

Reply:
If we provide the capability to override vmware default_run_post_script via cloud-init datasource config, we would need to document that override option in the ovf datasource docs (doc/rtd/topics/datasources/ovf.rst)  which gets published here https://cloudinit.readthedocs.io/en/latest/topics/datasources/ovf.html
Other datasources document their config options: [Azure](https://cloudinit.readthedocs.io/en/latest/topics/datasources/azure.html#configuration), [Ec2](https://cloudinit.readthedocs.io/en/latest/topics/datasources/ec2.html#configuration), [OpenStack](https://cloudinit.readthedocs.io/en/latest/topics/datasources/openstack.html#configuration)

Reply:
> @xiaofengw-vmware, sorry to rehash this earlier discussion, but I need a bit of clarification on the deployment 'flow' and a vmware user launching a vm in vCloud (not vmware admin) would follow in vCloud product.
> 
> I have a couple of questions about this architecture that hopefully you can clarift, because this doesn't seem to meet this concerns @smoser mentioned on the previous PR #426.
> 
> Please correct any misunderstanding I have about vCenter, vSphere, or vCloud products. If something sounds incorrect to you, it probably is due to my lack of understanding of VMware products.
> 
> 1. Is the intent of this branch is force the post-cloud-init-customization script to run of VM Tools configuration does not provide a value of true for "enable-custom-scripts" from the utility "vmware-toolbox-cmd". (I see the following [doc reference about enable-custom-scripts](https://kb.vmware.com/s/article/74880))
**[Xiaofengw-vmware]:** No. If "enable-custom-scripts" is set to true or false explicitly , then we will follow its value to enable/disable the script.  But if user doesn't set "enable-custom-scripts" explicitly, the default value of "enable-custom-scripts" is false. This fix only  wants to change the default value of  "enable-custom-scripts" to true  if "enable-cutom-scripts" is not set in VM Tools configuration. User still could disable the script by setting "enable-custom-scripts" to false.

> 2. Does a user launching each VM in vCloud have the opportunity to override this config option to "false" on the command line and/or UI during each vm creation of their images?  What does that look like in vCloud/vSphere world?
**[xiaofengw-vmware]:** Yes,  user could disable the script by setting "enable-custom-scripts" to false. Then even DEFAULT-RUN-POST-CUST-SCRIPT is set to true, the script still couldn't run. DEFAULT-RUN-POST-CUST-SCRIPT only works if "enable-custom-scripts" is not set.

> 3. I believe [the concern @smoser raised ](https://github.com/canonical/cloud-init/pull/426#issuecomment-644772401) may have been getting at is the following:
>    
>    * a given vCenter/vSphere or vCloud admin chooses a to default post-install script behavior to run on all VMs launched on that vCenter or vCloud deployment through VMTools configuration or through vCloud default_run_post_script setting
>    * it would be ideal if during each vm creation, the user launching an image/vm could specify an override in cloud-init which takes precedence over both vm tools config and vCloud default_run_post_script setting.
> 
> In this case we consider both VM Tools config and vCloud default_run_post_script to be "cloud platform default behavior" but we need a way for a cloud-init consumer to disable such behavior if desired.
> 
> One way to do that is to provide datasource-configuration options for DataSourceOVF which are read from the class attribute `self.ds_cfg` and is already being done for the flag [vmware_cust_file_max_wait](https://github.com/canonical/cloud-init/blob/master/cloudinit/sources/DataSourceOVF.py#L345-L364) . This would allow a customer to create a custom image that would override any default script opinions that a vCenter|vSphere/vCloud platform had chosen.
[xiaofengw-vmware]: I agree with you that we could add a such configuration in datasource-configuration. The problem is: vmware has its own  engine to initialize the VM,  by default cloud-init is disabled. After a VM / image is deployed, the vCenter or vCD administrator doesn't know whether cloud-init is enabled on VM or not.  He just collects the 
desired network setting / hostname and send this configuration to VM. Then VM Tools will check whether cloud-init is enabled. If no, vmware default engine will customize the VM per the configuration. Otherwise, DataSourceOVF of cloud-init will read meta data from the configuration and customize the VM following cloud-init work flow. So:
- We add this flag "DEFAULT-RUN-POST-CUST-SCRIPT" into the configuration, this flag could works for both vmware default engine and cloud-init.
- If we add this flag into datasource configuration, then we couldn't handle vmware default engine and cloud-init in the same way. Also, vCenter and vCD administrator need to know whether cloud-init or vmware default engine is enabled. Currently they don't need to know this.
- After this fix, the behavior will be:
   1) If VM administrator doesn't want others to execute a script on this VM, he could set "enable-custom-scripts" to false from the utility "vmware-toolbox-cmd".
   2) If VM administrator doesn't set value to "enable-custom-scripts", then by default this script is disabled for security purpose.
   3) For vCD product, they prefer to enable the script if "enable-custom-scripts" is not set. So vCD will generate a configuration file with "DEFAULT-RUN-POST-CUST-SCRIPT" set to true.  This flag works for both vmware customization engine and cloud-init.




Hi Chad, 
Thanks for the review, please see my answers in line.  Thanks for your time.

Reply:
@blackboxsw, could you take another look? Thanks.

Reply:
@blackboxsw @raharper Hi Chad, Ryan, please help to review my comments and check whether this design is ok. Thanks for your time.

Reply:
@xiaofengw-vmware sorry, just top-posted a response/question/suggestion two comments up

Reply:
> @xiaofengw-vmware sorry, just top-posted a response/question/suggestion two comments up

Thanks Chad for the detail comments, I think I have totally understood your concern.  As to the new switch "enable-custom-scripts" in datasourceOVF you suggested, I think it's a little duplicated with the one in VMTools configuration. The reason why I say that is that: The VMTools and datasourceOVF must work together.  The VMTools is help to fetch the meta data and custom script, then put them under a folder.  datasourceOVF start a loop to check that folder and apply the meta data/custom script once it find the meta data/custom script under that folder. For each cloud-init datasourceOVF user, he must install VMTools and be very clear with the VMTools configuration. I think it's a little confused if we add "enable-custom-script" for both VMTools configuration and datasourceOVF configuration.  I wonder whether we could setup a meeting to have a quick sync  and ensure there is no misunderstanding.  Please let me know when you are available. My timezone is  GMT+8 and I could attend the meeting in the morning or evening.  Thanks for your time in advance.

Reply:
Thanks @xiaofengw-vmware I think if this is a requirement of cloud-init support in VMWare, we should probably document that detail in a datasource document in docs at cloud-init/doc/rtd/topics/datasources/ovf.rst

It would be nice to have some semblance of OVF documentation for VMWare that states VMTools is required for cloud-init use of and maybe a list of the VMTools config properties that coud-init consumes.

This context could be added to a new Configuration section in the ovf.rst. An example of this format is
../cloud-init/doc/rtd/topics/datasources/ec2.rst

```
Configuration
-------------
On VMware platforms, VMtools use is required for OVF datasource configuration settings as well as vCloud and vSphere admin configuration. To override config settings when launching individual VMs from an ova, do the following....

The following VMTools configuration options affect cloud-init's behavior on a booted VM:
 * a 
 * b
 * c

For more information, see [this VMtools  documention](some-url)
```

Reply:
> Thanks @xiaofengw-vmware I think if this is a requirement of cloud-init support in VMWare, we should probably document that detail in a datasource document in docs at cloud-init/doc/rtd/topics/datasources/ovf.rst
> 
> It would be nice to have some semblance of OVF documentation for VMWare that states VMTools is required for cloud-init use of and maybe a list of the VMTools config properties that coud-init consumes.
> 
> This context could be added to a new Configuration section in the ovf.rst. An example of this format is
> ../cloud-init/doc/rtd/topics/datasources/ec2.rst
> 
> ```
> Configuration
> -------------
> On VMware platforms, VMtools use is required for OVF datasource configuration settings as well as vCloud and vSphere admin configuration. To override config settings when launching individual VMs from an ova, do the following....
> 
> The following VMTools configuration options affect cloud-init's behavior on a booted VM:
>  * a 
>  * b
>  * c 
> 
> VMWare admin can set the [config properties named in IMC config](https://github.com/canonical/cloud-init/blob/master/cloudinit/sources/helpers/vmware/imc/config.py)
> 
> For more information, see [this VMtools  documention](some-url) and specific VMTools parameters consumed
> ```
> 
> some thing like the above (obviously corrected as you have more details than we do) would give a little more context to how cloud-init interacts with VMware and VMTools.

@blackboxsw Thanks Chad, the configuration options are added into document, PTAL.